### DISCUSSION TITLE
Transposition not supported for `custom_jvp` with `pure_callback`

### DISCUSSION THREAD

**Discussion by antalszava** [jax-ml/jax#13277] [Q&A] [answered]
Hi all, I'm hitting the following error with an example that uses `custom_jvp` and `jax.pure_callback` to call `jax.jacrev`:
> ValueError: Pure callbacks do not support transpose. Please use `jax.custom_vjp` to use callbacks while taking gradients.

Any thoughts on when transpose may be supported for pure callbacks? Would there be any blockers to doing so, or is it more so a TODO for JAX?

Thank you!

**Comment by jakevdp** [comment, marked as answer]
JAX will not be able to automatically differentiate a pure callback without the user telling it what to do (pure_callback is essentially a black box to JAX's autodiff framework). The good news is you can use `jax.custom_jvp` to define the gradient for a function based on a pure callback. Here's a quick example that calls `np.sin` and `np.cos` on the host using a pure callback, and defines custom JVPs in terms of those callbacks:
```python
import jax
import numpy as np
jax.config.update('jax_enable_x64', True)

@jax.custom_jvp
def np_sin(x):
  return jax.pure_callback(np.sin, jax.ShapeDtypeStruct(np.shape(x), np.float64), x)

@np_sin.defjvp
def _np_sin_jvp(primals, tangents):
  x, = primals
  dx, = tangents
  return np_sin(x), np_cos(x) * dx  #d sin(x) = cos(x) dx

@jax.custom_jvp
def np_cos(x):
  return jax.pure_callback(np.cos, jax.ShapeDtypeStruct(np.shape(x), np.float64), x)

@np_cos.defjvp
def _np_cos_jvp(primals, tangents):
  x, = primals
  dx, = tangents
  return np_cos(x), -np_sin(x) * dx  # d cos(x) = -sin(x) dx

print(np_sin(1.0))
# 0.8414709848078965
print(jax.grad(np_sin)(1.0))
# 0.5403023058681398
print(jax.hessian(np_sin)(1.0))
# -0.8414709848078965
```
Note that `jax.hessian` is a second-order derivative that composes forward- and reverse-mode autodiff, showing that this is all JAX needs to correctly incorporate `pure_callback` into its autodiff system.

**Reply by antalszava** [reply]
Thank you! What if the functions that have a callback (`np_sin` & `np_cos`) use `tangents` too?

**Reply by jakevdp** [reply]
Can you elaborate? I'm not certain what you're asking.

**Reply by antalszava** [reply]
(Managed to surpass this issue, but keen to get more info for learning purposes.)

What if e.g., we had a `d_cos` function and it also used the `dx` coming from `tangents`? Is that feasible?
```py
@jax.custom_jvp
def d_cos(x, dx):
  return jax.pure_callback(lambda x, dx: np.sin(x) * dx, jax.ShapeDtypeStruct(np.shape(x), np.float64), x, dx)
```

Another minor question: why is `primals` and `tangents` one-element tuple that has to be unpacked?
```py
  x, = primals
  dx, = tangents
```

**Reply by jakevdp** [reply]
> What if e.g., we had a d_cos function and it also used the dx coming from tangents? Is that feasible?

"primals" and "tangents" only have special meanings within a JVP rule. In your `d_cos` function, you just have a function of two arguments, and those arguments can be anything. A jvp rule could happily pass any values as arguments to `d_cos`: it doesn't matter whether those values happen to be primals or tangents within the context of the JVP rule.

> Another minor question: why is primals and tangents one-element tuple that has to be unpacked?

Because `custom_jvp` is built to support functions that take any number of arguments, and by convention those arguments packaged as a tuple. For a function of N arguments, your primals and tangents are tuples of length-N. When N=1, you get a tuple of length-1.

**Reply by antalszava** [reply]
@jakevdp thank you! With an updated `np_cos` I'm getting an error now, what could be going wrong?
```python
import jax
import numpy as np
jax.config.update('jax_enable_x64', True)

@jax.custom_jvp
def np_sin(x):
    return jax.pure_callback(np.sin, jax.ShapeDtypeStruct(np.shape(x), np.float64), x)

@np_sin.defjvp
def _np_sin_jvp(primals, tangents):
    x, = primals
    dx, = tangents
    return np_sin(x), np_cos(x, dx) #d sin(x) = cos(x) dx

@jax.custom_jvp
def np_cos(x, dx=None):
    if dx is None:
        return jax.pure_callback(np.cos, jax.ShapeDtypeStruct(np.shape(x), np.float64), x)
    return jax.pure_callback(lambda x: np.cos(x) * dx, jax.ShapeDtypeStruct(np.shape(x), np.float64), x)

@np_cos.defjvp
def _np_cos_jvp(primals, tangents):
    x, = primals
    dx, = tangents
    return np_cos(x), -np_sin(x) * dx  # d cos(x) = -sin(x) dx

print(np_sin(1.0))
# 0.8414709848078965
print(jax.grad(np_sin)(1.0))
~/anaconda3/lib/python3.8/site-packages/jax/_src/callback.py in pure_callback(callback, result_shape_dtypes, vectorized, *args, **kwargs)
    129       lambda x: core.ShapedArray(x.shape, x.dtype), result_shape_dtypes)
    130   flat_result_avals, out_tree = tree_util.tree_flatten(result_avals)
--> 131   out_flat = pure_callback_p.bind(
    132       *flat_args, callback=_flat_callback,
    133       result_avals=tuple(flat_result_avals), vectorized=vectorized)

    [... skipping hidden 9 frame]

~/anaconda3/lib/python3.8/site-packages/jax/core.py in __array__(self, *args, **kw)
    539 
    540   def __array__(self, *args, **kw):
--> 541     raise TracerArrayConversionError(self)
    542 
    543   def __dlpack__(self, *args, **kw):

TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ShapedArray(float64[]):JaxprTrace(level=1/0)>
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerArrayConversionError
```

**Reply by jakevdp** [reply]
The issue is with this line:
```python
return jax.pure_callback(lambda x: np.cos(x) * dx, jax.ShapeDtypeStruct(np.shape(x), np.float64), x)
```
Your `pure_callback` function closes over `dx`, which is a traced value.

I'm a bit confused why your `np_cos` function takes a `dx` argument, but perhaps what you meant is something like this?
```python
return jax.pure_callback(lambda x, dx: np.cos(x) * dx, jax.ShapeDtypeStruct(np.shape(x), np.float64), x, dx)
```

**Reply by antalszava** [reply]
Thank you! Indeed, the second example you show is the one I had in mind.

Could you explain what you mean by
> pure_callback function closes over dx

? Does the `pure_callback` function not close over `x`?

Would returning `dx` "naively" help here?
```py
return jax.pure_callback(lambda x, dx: (np.cos(x) * dx, dx), jax.ShapeDtypeStruct(np.shape(x), np.float64), jax.ShapeDtypeStruct(np.shape(dx), np.float64), x, dx)
```
